File: nestMicro/clinical-micro/src/modules/tumorType/controllers/tumorType.controller.ts
import {Controller, Get, Post, Body, Param, ParseIntPipe, HttpCode, HttpStatus} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiParam, ApiOkResponse } from '@nestjs/swagger';
import { TumorTypeService } from '../services/tumorType.service';
import { CreateTumorTypeDto } from 'src/modules/tumorType/dto/create-tumorType';
import { TumorType } from 'src/modules/tumorType/entities/TumorType.entity';
import { stat } from 'fs';

@Controller('tumor-types')
export class TumorTypeController{
    constructor(private readonly tumorTypeService: TumorTypeService){}

    @Post()
    @HttpCode(HttpStatus.CREATED)
    @ApiOperation({
        summary: 'Registra un nuevo tipo de tumor',
        description: 'Crea una nueva clasificacion histológica con el sistema afectado.'
    })
    @ApiResponse({
        status: 201,
        description: 'El tipo de tumos ha sido creado exitosamente',
        type: TumorType
    })
    @ApiResponse({
        status: 400,
        description: 'Datos de entrada invalidos'
    })
    create(@Body() createTumorTypeDto: CreateTumorTypeDto): Promise <TumorType>{
        return this.tumorTypeService.create(createTumorTypeDto);
    }

    @Get()
    @ApiOperation({
        summary: 'Obtiene todos los tipos de tumor',
        description: ' Devuelve una lista completa de todas las clasificaciones de tumores registradas'
    })
    @ApiResponse({
        status: 200,
        description: 'Lista de tipos de tumores retornada exitosamente',
        type: [TumorType]
    })
    findAll(): Promise<TumorType[]>{
        return this.tumorTypeService.findAll();
    }

    @Get(':id')
    @ApiOperation({
        summary: 'Obtiene un tipo de tumor por ID',
        description: ' Busca y devuelve la clasificacion de tumor correspondiente al ID proporcionado'
    })
    @ApiParam({
        name: 'id',
        description: 'ID numerico del tipo de tumor',
        type: Number
    })
    @ApiResponse({
        status: 200,
        description: 'Tipo de tumor encontrado exitosamente.',
        type: TumorType
    })

    @ApiResponse({
        status: 404,
        description:'Tipo de tumor no encontrado'
    })
    findById(@Param('id', ParseIntPipe) id: number): Promise<TumorType>{
        return this.tumorTypeService.findById(id);
    }
<<<<<<< Updated upstream
=======

    @Patch(':id')
    @ApiOperation({
        summary: 'Actualizacion parcial de un tipo de tumor',
        description: 'Actualiza el nombre o el sistema afectado de un tipo de tumor existente.'
    })

    @ApiParam({
        name: 'id del tipo de tumor y DTO con los campos a actualizar',
        description: 'ID numerico del tipo de tumor',
        type: UpdateTumorTypeDto
    })

    @ApiResponse({
        status:200,
        description: 'Tipo de tumor actualizado exitosamente',
        type: TumorType
    })

    @ApiResponse({
        status:404,
        description: 'Tipo de tumor no encontrado'
    })

    updatePartial(
        @Param('id', ParseIntPipe) id:number,
        @Body()updateDto: UpdateTumorTypeDto): Promise<TumorType>{
            return this.tumorTypeService.updatePartial(id, updateDto);
        
    }

    @Delete(':id')
    @HttpCode(HttpStatus.NO_CONTENT)
    @ApiOperation({
        summary: 'Elimina un tipo de tumor',
        description: 'Elimina permanentemente un tipo de tumor por su ID'
    })

    @ApiParam({
        name: 'id',
        description: 'ID numerico del tipo de tumor a eliminar',
        type:Number
    })

    @ApiResponse({
        status: 204,
        description: 'Tipo de tumor no encontrado'
    })
    delete(@Param('id', ParseIntPipe) id: number): Promise<void>{
        return this.tumorTypeService.delete(id);
    }
>>>>>>> Stashed changes
    
}

File: springMicro/src/main/java/com/geno/springGateway/tumorType/infraestructure/controller/TumorTypeController.java
package com.geno.springGateway.tumorType.infraestructure.controller;
import com.geno.springGateway.common.ApiRestTemplateResponse;
import com.geno.springGateway.tumorType.application.dto.*;
import com.geno.springGateway.tumorType.application.service.TumorTypeService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.net.URI;
import org.springframework.web.util.UriComponentsBuilder;
import java.util.List;
/**
 * @author mendez
 */
@RestController
@RequestMapping("/tumor-types")
@RequiredArgsConstructor
@Tag(name = "Tipos de Tumor", description = "Operaciones CRUD sobre Tipos de Tumor (Proxy a NestJS)")
public class TumorTypeController {

    private final TumorTypeService tumorTypeService;


    @Operation(summary = "Obtener lista", description = "Recupera todos los Tipos de Tumor.")
    @GetMapping
    public ResponseEntity<ApiRestTemplateResponse<List<TumorTypeOutDto>>> getAll() {
        List<TumorTypeOutDto> tumors = tumorTypeService.findAll();
        return ResponseEntity.ok(new ApiRestTemplateResponse<>("Success", "Tipos de tumor encontrados", tumors));
    }

    @Operation(summary = "Obtener por ID", description = "Busca un Tipo de Tumor específico.")
    @ApiResponse(responseCode = "200", description = "Tipo de tumor encontrado con éxito.")
    @ApiResponse(responseCode = "404", description = "Tipo de tumor no encontrado (ID no existe).")
    @GetMapping("/{id}")
    public ResponseEntity<ApiRestTemplateResponse<TumorTypeOutDto>> getById(@PathVariable Long id) {
        TumorTypeOutDto dto = tumorTypeService.findById(id);
        return ResponseEntity.ok(new ApiRestTemplateResponse<>("Success", "Tipo de tumor encontrado", dto));
    }

    @Operation(summary = "Crear nuevo", description = "Registra un nuevo Tipo de Tumor.")
    @ApiResponse(responseCode = "201", description = "Tipo de tumor creado exitosamente.")
    @ApiResponse(responseCode = "400", description = "Solicitud inválida (error de validación en los datos de entrada).")
    @PostMapping
    public ResponseEntity<TumorTypeOutDto> create(@RequestBody TumorTypeInDto dto, UriComponentsBuilder ucb) {
        TumorTypeOutDto created = tumorTypeService.create(dto);
        URI locationUri = ucb.path("/tumor-types/{id}")
                .buildAndExpand(created.getId())
                .toUri();
        return ResponseEntity
                .created(locationUri)
                .body(created);
    }


    @Operation(summary = "Actualizar Nombre (Parcial)", description = "Actualiza solo el campo 'name' del Tipo de Tumor.")
    @ApiResponse(responseCode = "200", description = "Nombre de tipo de tumor actualizado con éxito.")
    @ApiResponse(responseCode = "400", description = "Solicitud inválida (el nombre está vacío o es nulo).")
    @ApiResponse(responseCode = "404", description = "Tipo de tumor no encontrado (ID no existe).")
    @PatchMapping("/{id}/name")
    public ResponseEntity<ApiRestTemplateResponse<TumorTypeOutDto>> patchName( @PathVariable Long id, @RequestBody UpdateTumorTypeNameDto dto)
    {
        TumorTypeOutDto patched = tumorTypeService.patchName(id, dto);
        return ResponseEntity.ok(new ApiRestTemplateResponse<>("Success", "Nombre de tipo de tumor actualizado", patched));
    }

    @Operation(summary = "Actualizar Sistema Afectado (Parcial)", description = "Actualiza solo el campo 'systemAffected'.")

    @ApiResponse(responseCode = "200", description = "Sistema afectado actualizado con éxito.")
    @ApiResponse(responseCode = "400", description = "Solicitud inválida (el campo está vacío o es nulo).")
    @ApiResponse(responseCode = "404", description = "Tipo de tumor no encontrado (ID no existe).")

            @PatchMapping("/{id}/system-affected")
    public ResponseEntity<ApiRestTemplateResponse<TumorTypeOutDto>> patchSystemAffected(
            @PathVariable Long id,
            @RequestBody UpdateTumorTypeSystemAffectedDto dto)
    {
        TumorTypeOutDto patched = tumorTypeService.patchSystemAffected(id, dto);
        return ResponseEntity.ok(new ApiRestTemplateResponse<>("Success", "Sistema afectado actualizado", patched));
    }

    @Operation(summary = "Eliminar", description = "Elimina un Tipo de Tumor por su ID.")
            @ApiResponse(responseCode = "200", description = "Tipo de tumor eliminado con éxito.")
            @ApiResponse(responseCode = "404", description = "Tipo de tumor no encontrado (ID no existe).")

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiRestTemplateResponse<Void>> delete(@PathVariable Long id) {
        tumorTypeService.delete(id);
        return new ResponseEntity<>(new ApiRestTemplateResponse<>("Success", "Tipo de tumor eliminado", null), HttpStatus.OK);
    }

}